pytket/extensions/cutensornet/general_state/tensor_network_state.py

```diff
@@ -0,0 +1,466 @@
+
+    def update_gates(self, gates_update_map: dict[str, NDArray]) -> None:
+        """Updates gate unitaries in the tensor network state.
+
+        Args:
+            gates_update_map: Map from gate (Command) opgroup name to a corresponding
+             gate unitary.
+
+        Raises:
+            ValueError: If a gate's unitary is of a wrong size.
+        """
+        for gate_label, unitary in gates_update_map.items():
+            gate_id = self._mutable_gates_map[gate_label]
+            gate_n_qubits = math.log2(unitary.shape[0])
+            if not gate_n_qubits.is_integer():
+                raise ValueError(
+                    f"Gate {gate_label} unitary's number of rows is not a power of two."
+                )
+            gate_tensor = cp.asarray(unitary, dtype="complex128").reshape(
+                [2] * (2 * int(gate_n_qubits)), order="F"
+            )
+            cutn.state_update_tensor(
+                self._handle, self._state, gate_id, gate_tensor.data.ptr, 1
+            )
+
+    def configure(self, attributes: Optional[dict] = None) -> GeneralState:
+        """Configures tensor network state for future contraction.
+
+        Args:
+            attributes: A dict of cuTensorNet State attributes and their values.
+
+        Returns:
+            Self (to allow for chaining with other methods).
+        """
+        if attributes is None:
+            attributes = {"NUM_HYPER_SAMPLES": 8}
+        attribute_values = [val for val in attributes.values()]
+        attribute_lst = [
+            getattr(cutn.StateAttribute, attr) for attr in attributes.keys()
+        ]
+        for attr, val in zip(attribute_lst, attribute_values):
+            attr_dtype = cutn.state_get_attribute_dtype(attr)
+            attr_arr = np.asarray(val, dtype=attr_dtype)
+            cutn.state_configure(
+                self._handle,
+                self._state,
+                attr,
+                attr_arr.ctypes.data,
+                attr_arr.dtype.itemsize,
+            )
+        return self
+
+    def prepare(self, scratch_fraction: float = 0.5) -> GeneralState:
+        """Prepare tensor network state for future contraction.
+
+        Allocates workspace memory necessary for contraction.
+
+        Raises:
+            MemoryError: If there is insufficient workspace on GPU.
+
+        Args:
+            scratch_fraction: Fraction of free memory on GPU to allocate as scratch
+             space.
+
+        Returns:
+            Self (to allow for chaining with other methods).
+        """
+        self._stream = (
+            cp.cuda.Stream()
+        )  # In current cuTN release it is unused (could be 0x0)
+        free_mem = self._dev.mem_info[0]
+        scratch_size = int(scratch_fraction * free_mem)
+        self._scratch_space = cp.cuda.alloc(scratch_size)
+        self._logger.debug(f"Allocated {scratch_size} bytes of scratch memory on GPU")
+        self._work_desc = cutn.create_workspace_descriptor(self._handle)
```

Comment:
As pointed out in [the docs](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/functions.html#cutensornetexpectationcompute), in the current version of cuTensorNet, it is necessary to provide a workspace descriptor explicitly. In the upcoming version this will not be needed (as far as I can tell from conversations with them).

Reply:
Turns out that the only difference between the preparation and allocation for statevector and for expectation value is the line `state_prepare(` and `expectation_prepare(`. Unfortunately, it's in the middle of the boilerplate code and I've tried to move it around but it results in cuTensorNet errors... 

Considering that this scratch space allocation should no longer be necessary in the next version of cuTensorNet (which I gathered from conversations with the devs), I'm not willing to make this prettier: this code will be removed soon enough anyway.